File: core/report.py
"""Build and export a CSV report of a send run."""

from __future__ import annotations

import csv
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path


@dataclass
class SendResult:
    timestamp: str
    name: str
    phone: str
    status: str
    detail: str = ""

    @classmethod
    def now(cls, name: str, phone: str, status: str, detail: str = "") -> "SendResult":
        return cls(
            timestamp=datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            name=name,
            phone=phone,
            status=status,
            detail=detail,
        )


def export_results(results: list[SendResult], path: str | Path) -> None:
    """Write send results to a UTF-8 CSV (Excel-friendly, semicolon-separated)."""
    path = Path(path)
    with open(path, "w", encoding="utf-8-sig", newline="") as f:
        writer = csv.writer(f, delimiter=";")
        writer.writerow(["Tijdstip", "Naam", "Telefoonnummer", "Status", "Details"])
        for r in results:
            writer.writerow([r.timestamp, r.name, r.phone, r.status, r.detail])

File: apps/go/internal/report/report.go
package report

import (
	"encoding/csv"
	"fmt"
	"os"
	"time"
)

// SendResult is one row in the export CSV.
type SendResult struct {
	Timestamp string `json:"timestamp"`
	Name      string `json:"name"`
	Phone     string `json:"phone"`
	Status    string `json:"status"`
	Detail    string `json:"detail"`
}

// NewResult creates a result with the current timestamp.
func NewResult(name, phone, status, detail string) SendResult {
	return SendResult{
		Timestamp: time.Now().Format("2006-01-02 15:04:05"),
		Name:      name,
		Phone:     phone,
		Status:    status,
		Detail:    detail,
	}
}

// Export writes send results to a UTF-8-BOM semicolon-separated CSV (Excel-friendly).
func Export(results []SendResult, path string) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	// UTF-8 BOM
	if _, err := f.Write([]byte{0xEF, 0xBB, 0xBF}); err != nil {
		return err
	}

	w := csv.NewWriter(f)
	w.Comma = ';'
	if err := w.Write([]string{"Tijdstip", "Naam", "Telefoonnummer", "Status", "Details"}); err != nil {
		return err
	}
	for _, r := range results {
		if err := w.Write([]string{r.Timestamp, r.Name, r.Phone, r.Status, r.Detail}); err != nil {
			return err
		}
	}
	w.Flush()
	if err := w.Error(); err != nil {
		return fmt.Errorf("csv flush: %w", err)
	}
	return nil
}
